packages/api/lib/analytics.js

```diff
@@ -12,32 +12,73 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+const NUM_STORE = 15;
+
 class TestCaseAnalytics {
   // params: encoded name of a test and this.client.collection(global.headCollection).doc(repoid)
-  constructor (testCase, snapshot) {
+  constructor (testCase, cachedSuccess, cachedFails, buildInfo) {
+    this.cachedSuccess = cachedSuccess.map(x => x.toDate());
+    this.cachedFails = cachedFails.map(x => x.toDate());
     this.testCase = testCase;
-    this.snapshot = snapshot;
-    this.snapshotData = [];
-    this.snapshot.forEach((doc) => this.snapshotData.push(doc.data()));
+    this.buildInfo = buildInfo;
+    if (testCase.successful) {
+      this.addDateToList(this.cachedSuccess);
+    } else {
+      this.addDateToList(this.cachedFails);
+    }
+
+    this.createMergedList();
+
+    if (this.mergedList.length > NUM_STORE) {
+      this.deleteFirst();
+    }
   }
 
-  computePassingPercent () {
-    let successCnt = 0;
-    let total = 0;
-    this.snapshotData.forEach((doc) => {
-      successCnt += doc.status === 'OK' ? 1 : 0;
-      total += 1;
+  createMergedList () {
+    this.mergedList = [];
+    this.cachedFails.forEach(f => {
+      this.mergedList.push({
+        date: f,
+        passed: false
+      });
+    });
+    this.cachedSuccess.forEach(s => {
+      this.mergedList.push({
+        date: s,
+        passed: true
+      });
     });
-    return (total > 0) ? successCnt / parseFloat(total) : -1;
+
+    this.mergedList.sort(function (a, b) {
```

Comment:
by performing this merge in the API rather than in the database, we run the potential of losing data when two updates have happened in parallel -- this is very likely to happen in cases like flaky.dev where multiple test suites are running in parallel across a few matrices ... I wonder if this might be a potential root cause of #180.

I think denormalization is a brilliant idea, for the reasons you've outlined, but I wonder if we could move either to using firebase's [push](https://firebase.google.com/docs/database/admin/save-data#node.js_9) functionality, which would ensure that two parallel updates are both appended.

You'd then just need to add one one extra operation that deletes the oldest item, if and only if there are 15 or more.